tests/__init__.py

```diff
@@ -420,68 +428,118 @@ def start_xdp(self):
         """
 
+        portal_dir = (
+            Path(os.getenv("G_TEST_BUILDDIR") or "tests")
+            / "portals"
+            / "test"
+        )
+
+        if not portal_dir.exists():
+            raise FileNotFoundError(
+                f"{portal_dir} does not exist, try running from meson build dir or setting G_TEST_SRCDIR"
+            )
+
+        env = os.environ.copy()
+        env["G_DEBUG"] = "fatal-criticals"
+        env["XDG_DESKTOP_PORTAL_DIR"] = portal_dir
+        env["XDG_CURRENT_DESKTOP"] = "test"
+        env["XDG_DESKTOP_PORTAL_TEST_APP_ID"] = self.app_id
+
+        if self.umockdev:
+            env["UMOCKDEV_DIR"] = self.umockdev.get_root_dir()
+
         self.start_dbus_monitor()
+        self.start_portal_frontend(env)
+        self.start_permission_store(env)
 
+    def start_portal_frontend(self, env):
         # This roughly resembles test-portals.c and glib's test behavior
         # but preferences in-tree testing by running pytest in meson's
         # project_build_root
         libexecdir = os.getenv("LIBEXECDIR")
         if libexecdir:
-            xdp_path = Path(libexecdir) / "xdg-desktop-portal"
+            portal_frontend = Path(libexecdir) / "xdg-desktop-portal"
         else:
-            xdp_path = (
+            portal_frontend = (
                 Path(os.getenv("G_TEST_BUILDDIR") or "tests")
                 / ".."
                 / "src"
                 / "xdg-desktop-portal"
             )
 
-        if not xdp_path.exists():
+        if not portal_frontend.exists():
             raise FileNotFoundError(
-                f"{xdp_path} does not exist, try running from meson build dir or setting G_TEST_BUILDDIR"
+                f"{portal_frontend} does not exist, try running from meson build dir or setting G_TEST_BUILDDIR"
             )
 
-        portal_dir = Path(os.getenv("G_TEST_BUILDDIR") or "tests") / "portals" / "test"
-        if not portal_dir.exists():
+        portal_frontend = subprocess.Popen([portal_frontend], env=env)
+
+        for _ in range(50):
+            if self.dbus_test_case.dbus_con.name_has_owner("org.freedesktop.portal.Desktop"):
+                break
+            time.sleep(0.1)
+        else:
+            assert (
+                False
+            ), "Timeout while waiting for xdg-desktop-portal to claim the bus"
+
+        self.portal_frontend = portal_frontend
+
+    def start_permission_store(self, env):
+        """
+        Start the xdg-permission-store process
+        """
+
+        # This roughly resembles test-portals.c and glib's test behavior
+        # but preferences in-tree testing by running pytest in meson's
+        # project_build_root
+        libexecdir = os.getenv("LIBEXECDIR")
+        if libexecdir:
+            permission_store_path = Path(libexecdir) / "xdg-permission-store"
+        else:
+            permission_store_path = (
+                Path(os.getenv("G_TEST_BUILDDIR") or "tests")
```

Comment:
```suggestion
                Path(os.getenv("G_TEST_BUILDDIR", "tests"))
```
same as above but feel free to ignore, the `or` is more expressive